src/java-interop/java-interop-jvm.h

```diff
@@ -5,19 +5,14 @@
 
 typedef void JavaVM;
 
-typedef int (*java_interop_JNI_CreateJavaVM_fptr) (JavaVM **p_vm, void **p_env, void *vm_args);
-typedef int (*java_interop_JNI_GetCreatedJavaVMs_fptr) (JavaVM **vmBuf, int bufLen, int *nVMs);
-
-/* NOTE: structure members MUST NOT CHANGE ORDER. */
-struct DylibJVM {
-	void						       *dl_handle;
-	java_interop_JNI_CreateJavaVM_fptr			JNI_CreateJavaVM;
-	java_interop_JNI_GetCreatedJavaVMs_fptr			JNI_GetCreatedJavaVMs;
-};
-
 JAVA_INTEROP_BEGIN_DECLS
 
-MONO_API    int java_interop_jvm_load (const char *path);
+#define JAVA_INTEROP_JVM_FAILED_LOAD    (-1000)
+#define JAVA_INTEROP_JVM_FAILED_LOADED  (-1001)
+#define JAVA_INTEROP_JVM_FAILED_SYMBOL  (-1002)
+
+MONO_API    int java_interop_jvm_load_path (const char *path);
+MONO_API    int java_interop_jvm_load_handle (void *handle);
```

Comment:
Do we need to have that split into 2 functions? How would be java_interop_jvm_load_handle used on Android?

I expected to just pass NULL as path on Android and use RTLD_DEFAULT instead of calling dlopen? Like
```
int
java_interop_jvm_load (const char *path)
 {
	void *h = path ? dlopen (path, RTLD_LAZY) : RTLD_DEFAULT;
	if (!h) {
		return JAVA_INTEROP_JVM_FAILED_LOAD;
	}
	...
}
```
Or do we have the handle available in monodroid?

Reply:
On hosts we use the path. I am just not sure whether there is scenario, where we need to pass the handle directly. Android was mentioned, but I don't see where to get it, other than using `RTLD_DEFAULT` special handle, which we can get by using NULL path.

Using `void *h = dlopen (path, RTLD_LAZY);` is probably better than using the `RTLD_DEFAULT` directly (as I suggested earlier), because it will add `RTLD_LAZY` to the `RTLD_DEFAULT` handle.

Reply:
For some reason (insanity?) I thought that using `dlsym(NULL, "symbol")` would work, so `java_interop_jvm_load_handle(NULL)` would be a perfectly sensible thing to do.

I'm not sure what I was thinking.

That said, now that I read the [**dlsym**(3) man page](https://linux.die.net/man/3/dlsym), my crazy thinking doesn't seem *that* crazy, in that `java_interop_jvm_load_handle(RTLD_DEFAULT)` seems like a perfectly sane thing to do on Android.

Thus, having the two overloads seems...okay?  Though using `java_interop_jvm_load_path(NULL)` *also* seems like it should be okay...

...except on Windows, which IIRC doesn't have a decent equivalent to `dlopen(NULL, ...)`.  However, it likewise has no equivalent to `dlsym(RTLD_DEFAULT, ...)`, so for "proper" Windows use we'd actually need to use `java_interop_jvm_load_path("path/to/jvm.dll")` *anyway*, which removes the argument for having `java_interop_jvm_load_handle()`.

@grendello, @radekdoulik: In short, which native API makes most sense to you?

Reply:
I think `java_interop_jvm_load_path(NULL)` is preferable as it would still use the `RTLD_LAZY`, so it will be better performance wise.

Reply:
windows can use the OSS library we build in MXE, so you can use dlopen (we [submodule it in XA](https://github.com/dlfcn-win32/dlfcn-win32/tree/ef7e412d3fa2b511e660342e237536c235e51ee4)). As for the native API, `dlsym(RTLD_DEFAULT)` appears to be the one that's more expressive and descriptive.

Reply:
@grendello: I don’t understand your comment about `dlsym(RTLD_DEFAULT)`.

The question is what to do about `java_interop_jvm_load_path()` vs. `java_interop_jvm_load_handle()`. 

Supporting `dlsym(RTLD_DEFAULT)` could be done by providing both, as this PR does, or by special-casing `java_interop_jvm_load_path(NULL)` so that when `path==NULL` we do `handle=RTLD_DEFAULT`, or by assuming that `dlsym(dlopen(NULL, RTLD_LAZY), "symbol")` is equivalent to `dlsym(RTLD_DEFAULT, "symbol")`.

Reply:
ignore my comment, I misunderstood. I thought you were asking about libdl APIs to use. From the `java_interop*` ones, I think the one taking `char*` is the right one to use. It can be used with the `dlfcn-win32` library on the Windows side without any problems.